security/pkg/k8s/chiron/utils.go

```diff
@@ -215,150 +228,118 @@ func reloadCACert(wc *WebhookController) (bool, error) {
 	return certChanged, nil
 }
 
-func submitCSR(certClient certclient.CertificateSigningRequestInterface,
-	csrName string,
-	csrSpec *cert.CertificateSigningRequestSpec,
-	numRetries int) (*cert.CertificateSigningRequest, error) {
-	k8sCSR := &cert.CertificateSigningRequest{
-		TypeMeta: metav1.TypeMeta{
-			APIVersion: "certificates.k8s.io/v1beta1",
-			Kind:       "CertificateSigningRequest",
-		},
-		ObjectMeta: metav1.ObjectMeta{
-			Name: csrName,
-		},
-		Spec: *csrSpec,
-	}
-	var reqRet *cert.CertificateSigningRequest
-	var errRet error
+func submitCSR(clientset clientset.Interface,
+	csrData []byte, csrName string, signerName string,
+	usages []certv1.KeyUsage, numRetries int) (*certv1.CertificateSigningRequest, *certv1beta1.CertificateSigningRequest, error) {
+	var err error = fmt.Errorf("unable to submit csr %s", csrName)
+
 	for i := 0; i < numRetries; i++ {
 		log.Debugf("trial %v to create CSR (%v)", i, csrName)
-		reqRet, errRet = certClient.Create(context.TODO(), k8sCSR, metav1.CreateOptions{})
-		if errRet == nil && reqRet != nil {
-			break
+		csr := &certv1.CertificateSigningRequest{
+			// Username, UID, Groups will be injected by API server.
+			TypeMeta: metav1.TypeMeta{Kind: "CertificateSigningRequest"},
+			ObjectMeta: metav1.ObjectMeta{
+				Name: csrName,
+			},
+			Spec: certv1.CertificateSigningRequestSpec{
+				Request:    csrData,
+				Usages:     usages,
+				SignerName: signerName,
+			},
 		}
-		// If an err other than the CSR exists is returned, re-try
-		if !kerrors.IsAlreadyExists(errRet) {
-			log.Debugf("failed to create CSR (%v): %v", csrName, errRet)
-			continue
+		if len(csr.Spec.Usages) > 0 && len(csr.Spec.SignerName) > 0 && csr.Spec.SignerName != "kubernetes.io/legacy-unknown" {
```

Comment:
why usages need to be checked

Reply:
Was borrowing the logic from here https://github.com/kubernetes/client-go/blob/master/util/certificate/csr/csr.go. Prefer to use logic already committed (and reviewed) in the kubernetes codebase. Can't use this package directly because it has some missing functionality (for approval/deletion of kubernetes csr requests)

Reply:
agreed